### PR TITLE
WIP: Fix CPP mangling #1

### PR DESCRIPTION
**Pull request by Darredevil**
While trying to interface D with C++ I noticed that most cases will not work due to bugs in the name mangling which causes linking errors. One big problem was the symbol substitution, first of all the way the dmd checked if a symbol was already present in the substitution table was broken for certain cases. The second problem was the way the table was created (order of insertion is relevant here).

For example `std::pair<void*, void*>::swap(std::pair<void*, void*>&)` was mangle as:

-  __ZNSt4pairIPvS0_E4swapERS1_ (gcc)

-  _ZNSt4pairIPvS1_E4swapERStS0_IS1_S1_E (dmd)

Now dmd mangles these cases the same way as gcc. Further tests and fixes are required because currently 

> No attempt is made to support mangling of templates, operator overloading, or special functions

Will add future patches as I stumble upon the issues and fix them.

### CODE REVIEW

**Comment by WalterBright** [comment]
This is good but it must have a test case. Add one to `test/runnable/cppa.d`. Please ensure that all the changed lines are covered by the test.

**Comment by ibuclaw** [comment]
Storing should be after the symbol has been written to buffer, there should be no problem there.

Not so sure about the strcmp however.  There's probably a very good reason why they don't match.

**Comment by ibuclaw** [comment]
> No attempt is made to support mangling of templates, operator overloading, or special functions

And I'm not so sure any attempt should be made to either.  With exception to operators which gets a long, drawn out maybe.

**Comment by andralex** [comment]
I'm okay with this with @WalterBright's amendment and of course pending pass of tests that currently all fail.

**Review by andralex — APPROVED** [review]
Please add a test (can be only D code using `extern(C++)`) and make sure all tests pass.

**Comment by ibuclaw** [comment]
> can be only D code using extern(C++)

No, anything marked extern(C++) should be calling a function implemented in C++, otherwise what's the point?

**Comment by andralex** [comment]
@Darredevil ping

**Comment by Darredevil** [comment]
So far, when I fix a test or particular case it seems another breaks.
Will tackle them as they pop up until all the tests pass, including the new ones.

File: src/ddmd/cppmangle.d [hunk 1]
```diff
@@ -263,13 +276,13 @@ static if (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TAR
                 }
                 if (p && !p.isModule())
                 {
-                    if (p.ident == Id.std && is_initial_qualifier(p))
+                    if (p.ident == Id.std && is_initial_qualifier(p)) {
```
Comment (andralex) [inline src/ddmd/cppmangle.d]: no Egyptian bracing please - put brace on its own line

File: src/ddmd/cppmangle.d [hunk 2]
```diff
@@ -263,13 +276,13 @@ static if (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TAR
...
                         buf.writestring("St");
-                    else
+                    }
+                    else {
```
Comment (andralex) [inline src/ddmd/cppmangle.d]: ditto

File: src/ddmd/cppmangle.d [hunk 3]
```diff
@@ -433,8 +444,9 @@ static if (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TAR
             if (p && !p.isModule() && tf.linkage == LINKcpp && !ftd)
             {
                 buf.writeByte('N');
-                if (d.type.isConst())
+                if (d.type.isConst()) {
```
Comment (andralex) [inline src/ddmd/cppmangle.d]: throughout

**Review by andralex — CHANGES_REQUESTED** [review]
Tests are still failing. In order to improve development speed, you may want to run tests locally before pushing code.

File: src/ddmd/cppmangle.d (hunk 1 above)
Comment (Darredevil) [inline src/ddmd/cppmangle.d]: Ignore this for now please, it will be squashed and fixed in future commit

File: src/ddmd/cppmangle.d [hunk 4]
```diff
@@ -486,6 +501,7 @@ static if (TARGET_LINUX || TARGET_OSX || TARGET_FREEBSD || TARGET_OPENBSD || TAR
                 assert(tf.ty == Tfunction);
                 argsCppMangle(tf.parameters, tf.varargs);
             }
+
```
Comment (RazvanN7) [inline src/ddmd/cppmangle.d]: No need for this empty line

**Pull request closed without merging**